### PR TITLE
#69 Fix issue with inheritance in aql.render

### PR DESCRIPTION
**Pull request by dimberman**
There were three primary issues with aql.render where the DAGs were
showing up in the graph view properly but were not running properly when
tasks inherit from eachother.

1. We were not properly returning the result which meant that the Table
   object was not being saved to XCOM
2. (snowflake specific) There was an issue in how we would parse
   snowflake parameters that prevented them from properly overwriting
   parameters in the SQL statement.
3. There was a bug in snowflake when attempting to handle multiple
   schemas at the same time.

### CODE REVIEW

File: src/astro/utils/snowflake_transform.py [hunk 1]
```diff
@@ -20,7 +20,11 @@
 
 
 def handle_table(t: Table):
-    return t.schema + "." + t.table_name if t.schema else t.table_name
+    return (
+        t.database + "." + t.schema + "." + t.table_name
+        if t.schema and "." not in t.table_name
+        else t.table_name
```
Comment (tatiana) [inline src/astro/utils/snowflake_transform.py:26]: In which circumstances do we expect `table_name` to contain  `.`?

File: tests/operators/test_snowflake_decorator.py [hunk 2]
```diff
@@ -142,7 +152,7 @@ def tearDown(self):
     def test_snowflake_query(self):
         @aql.transform
         def sample_snow(input_table: Table):
-            return f"SELECT * FROM {input_table.table_name} LIMIT 10"
+            return "SELECT * FROM {input_table} LIMIT 10"
```
Comment (tatiana) [inline tests/operators/test_snowflake_decorator.py:155]: Why don't we need the `f-string` anymore?

File: tests/parsers/test_sql_directory_parser.py [hunk 3]
```diff
@@ -98,3 +98,10 @@ def test_parse_frontmatter(self):
             new_customers_table.operator.sql
             == "SELECT * FROM {customers_table} WHERE member_since > DATEADD(day, -7, '{{ execution_date }}')"
         )
+
+    def test_parse_creates_xcom(self):
+        with self.dag:
+            rendered_tasks = aql.render(dir_path + "/single_task_dag")
+        from tests.operators import utils as test_utils
```
Comment (tatiana) [inline tests/parsers/test_sql_directory_parser.py:105]: What's the benefit in having the import statement from within the test function?

File: tests/parsers/test_sql_directory_parser.py [hunk 4]
```diff
@@ -98,3 +98,10 @@ def test_parse_frontmatter(self):
...
+
+        test_utils.run_dag(self.dag)
```
Comment (tatiana) [inline tests/parsers/test_sql_directory_parser.py:107]: How are we confirming the table object was recorded in XCOM?
Are we missing any asserts in this test?

File: src/astro/sql/operators/agnostic_boolean_check.py [hunk 5]
```diff
@@ -120,7 +120,7 @@ def execute(self, context: Dict):
         self.handler = lambda curr: curr.fetchall()
 
         self.sql, self.parameters = execute_boolean_checks(
-            self.table.table_name, self.checks, self.conn_id  # type: ignore
+            self.table, self.checks, self.conn_id  # type: ignore
```
Comment (tatiana) [inline src/astro/sql/operators/agnostic_boolean_check.py:123]: The docstring in lines 68-69 also needs to be updated

File: tests/operators/test_snowflake_decorator.py [hunk 6]
```diff
@@ -166,10 +176,20 @@ def sample_snow(input_table: Table):
...
+                    schema=os.getenv("SNOWFLAKE_SCHEMA"),
+                    database=os.getenv("SNOWFLAKE_DATABASE"),
+                    warehouse=os.getenv("SNOWFLAKE_WAREHOUSE"),
+                ),
+            )
         test_utils.run_dag(self.dag)
 
         df = hook.get_pandas_df(
-            f'SELECT * FROM "{os.getenv("SNOWFLAKE_DATABASE")}"."{os.getenv("SNOWFLAKE_SCHEMA")}"."{self.snowflake_table}"'
+            f'SELECT * FROM "{os.getenv("SNOWFLAKE_DATABASE")}"."{os.getenv("SNOWFLAKE_SCHEMA")}"."{self.snow_inherit_table}"'
```
Comment (tatiana) [inline tests/operators/test_snowflake_decorator.py:192]: Do we still have other tests covering the non-inheritance, more basic, use case?

**Review by tatiana — COMMENTED** [review]
Thank you very much for identifying and fixing all those issues at the speed of light, @dimberman ! :rocket: 

Some questions inline

File: src/astro/sql/operators/sql_decorator.py [hunk 7]
```diff
@@ -166,6 +166,10 @@ def handle_schema(self, output_table_name):
         """
         if self.conn_type == "postgres" and self.schema:
             output_table_name = self.schema + "." + output_table_name
+        elif self.conn_type == "snowflake" and self.schema and "." not in self.sql:
```
Comment (utkarsharma2) [inline src/astro/sql/operators/sql_decorator.py:169]: Why are we checking based on  - "." not in self.sql?

File: src/astro/utils/snowflake_append.py [hunk 8]
```diff
@@ -34,11 +34,11 @@ def wrap_identifier(inp):
 
     if columns or casted_columns:
         statement = (
-            "INSERT INTO Identifier(%(main_table)s) ({main_cols}{sep}{main_casted_cols})"
-            "(SELECT {fields}{sep}{casted_fields} FROM Identifier(%(append_table)s))"
+            "INSERT INTO %(main_table)s ({main_cols}{sep}{main_casted_cols})"
```
Comment (utkarsharma2) [inline src/astro/utils/snowflake_append.py:37]: why are we removing - Identifier()? this may lead to SQL injections right?

File: src/astro/utils/snowflake_transform.py (hunk 1 above)
Comment (dimberman) [reply src/astro/utils/snowflake_transform.py:26]: @tatiana if the user provides a full tablename path. I've seen multiple test users in snowflake choose to identify the entire table name "DATABASE"."SCHEMA"."TABLE NAME"

File: tests/operators/test_snowflake_decorator.py (hunk 2 above)
Comment (dimberman) [reply tests/operators/test_snowflake_decorator.py:155]: @tatiana these were never supposed to be f-strings and by identifying them as f-strings users are actually opening themselves up to SQL injections. This raises a huge question over whether our current system looks too much like f-strings and therefore might promote bad habits.

File: tests/parsers/test_sql_directory_parser.py (hunk 3 above)
Comment (dimberman) [reply tests/parsers/test_sql_directory_parser.py:105]: None I can move that

File: src/astro/sql/operators/sql_decorator.py (hunk 7 above)
Comment (dimberman) [reply src/astro/sql/operators/sql_decorator.py:169]: Because if the user has already identified the schema and database in the table_name then we don't need to add it.

see here: https://github.com/astro-projects/astro/pull/68#discussion_r799533367

File: src/astro/utils/snowflake_append.py (hunk 8 above)
Comment (dimberman) [reply src/astro/utils/snowflake_append.py:37]: The identifier is actually now being added in in the sql_decorator.py. We now have it set that it can detect tables and automatically do that. We don't have that same detection for columns which is why those are still there.

File: src/astro/utils/snowflake_transform.py (hunk 1 above)
Comment (dimberman) [reply src/astro/utils/snowflake_transform.py:32]: @tatiana I added a docstring and made it more explicit

Comment (dimberman) [reply src/astro/utils/snowflake_transform.py]: Actually let's just always do fully qualified name. I'm worried there will be edge cases with table names that have period in them.

File: tests/parsers/test_sql_directory_parser.py (hunk 4 above)
Comment (tatiana) [reply tests/parsers/test_sql_directory_parser.py:111]: @dimberman I would feel more comfortable if the test confirmed that one task depends on the other and had some more explicit checks. If someone inadvertently changed the files within `dir_path + "/single_task_dag"`, this test would likely continue working - but we would not be testing the original problem.
